File: parallel-programming-with-ms-dotnet/CSharp/Chapter5/a-Dash/AnalysisEngine.cs
//===============================================================================
// Microsoft patterns & practices
// Parallel Programming Guide
//===============================================================================
// Copyright © Microsoft Corporation.  All rights reserved.
// This code released under the terms of the 
// Microsoft patterns & practices license (http://parallelpatterns.codeplex.com/license).
//===============================================================================

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Practices.ParallelGuideSamples.ADash.BusinessObjects;
using Microsoft.Practices.ParallelGuideSamples.Utilities;

namespace Microsoft.Practices.ParallelGuideSamples.ADash
{
    /// <summary>
    /// Component for data analysis
    /// </summary>
    public class AnalysisEngine : IAnalysisEngine
    {
        // internal scaling factor
        readonly double speedFactor;

        CancellationTokenSource cts = null;

        public AnalysisEngine()
            : this(1.0d)
        {
        }

        public AnalysisEngine(double speed)
        {
            speedFactor = speed;
        }

        #region CreateSampleData

        static IList<StockData> MakeNyseSecurityInfo()
        {
            return GenerateSecurities("NYSE", 100);
        }

        static IList<StockData> MakeNasdaqSecurityInfo()
        {
            return GenerateSecurities("NASDAQ", 100);
        }

        static IList<StockData> MakeFedSecurityInfo()
        {
            return GenerateSecurities("", 100);
        }

        static IList<StockData> GenerateSecurities(string exchange, int size)
        {
            var result = new List<StockData>();
            for (int i = 0; i < size; i++)
                result.Add(new StockData(exchange + " Stock " + i, new[] { 0.0d, 1.0d, 2.0d }));
            return result;
        }

        #endregion

        #region Analysis Helper Methods

        StockDataCollection LoadNyseData()
        {
            SampleUtilities.DoIoIntensiveOperation(2.5, cts.Token);
            if (cts.Token.IsCancellationRequested)
                return null;
            return new StockDataCollection(MakeNyseSecurityInfo());
        }

        StockDataCollection LoadNasdaqData()
        {
            SampleUtilities.DoIoIntensiveOperation(2.0 * speedFactor, cts.Token);
            if (cts.Token.IsCancellationRequested)
                return null;

            return new StockDataCollection(MakeNasdaqSecurityInfo());
        }

        StockDataCollection LoadFedHistoricalData()
        {
            SampleUtilities.DoIoIntensiveOperation(3.0 * speedFactor, cts.Token);
            if (cts.Token.IsCancellationRequested)
                return null;

            return new StockDataCollection(MakeFedSecurityInfo());
        }

        StockDataCollection MergeMarketData(IEnumerable<StockDataCollection> allMarketData)
        {
            SampleUtilities.DoCpuIntensiveOperation(2.0 * speedFactor, cts.Token);
            var securities = new List<StockData>();

            if (!cts.Token.IsCancellationRequested)
            {
                foreach (StockDataCollection md in allMarketData)
                    securities.AddRange(md);
            }

            if (cts.Token.IsCancellationRequested)
                return null;
            else
                return new StockDataCollection(securities);
        }

        /// <summary>
        /// Normalize stock data.
        /// </summary>
        StockDataCollection NormalizeData(StockDataCollection marketData)
        {
            SampleUtilities.DoCpuIntensiveOperation(2.0 * speedFactor, cts.Token);
            if (cts.Token.IsCancellationRequested)
                return null;
            else
                return new StockDataCollection(marketData);
        }

        StockAnalysisCollection AnalyzeData(StockDataCollection data)
        {
            if (cts.Token.IsCancellationRequested)
                return null;
            return MarketAnalyzer.Run(data);
        }

        MarketModel RunModel(StockAnalysisCollection data)
        {
            SampleUtilities.DoCpuIntensiveOperation(2.0 * speedFactor, cts.Token);
            if (cts.Token.IsCancellationRequested)
                return null;
            else
                return MarketModeler.Run(data);
        }

        MarketRecommendation CompareModels(IEnumerable<MarketModel> models)
        {
            SampleUtilities.DoCpuIntensiveOperation(2.0 * speedFactor, cts.Token);
            if (cts.Token.IsCancellationRequested)
                return null;
            else
                return ModelComparer.Run(models.ToArray());
        }

        #endregion

        #region Analysis Public Methods

        /// <summary>
        /// Creates a market recommendation using a fully sequential operation
        /// </summary>
        /// <returns>A market recommendation</returns>
        public MarketRecommendation DoAnalysisSequential()
        {
            StockDataCollection nyseData = 
                LoadNyseData();
            StockDataCollection nasdaqData = 
                LoadNasdaqData();
            StockDataCollection mergedMarketData = 
                MergeMarketData(new[]{nyseData, nasdaqData});
            StockDataCollection normalizedMarketData = 
                NormalizeData(mergedMarketData);
            StockDataCollection fedHistoricalData = 
                LoadFedHistoricalData();
            StockDataCollection normalizedHistoricalData = 
                NormalizeData(fedHistoricalData);
            StockAnalysisCollection analyzedStockData = 
                AnalyzeData(normalizedMarketData);
            MarketModel modeledMarketData = 
                RunModel(analyzedStockData);
            StockAnalysisCollection analyzedHistoricalData = 
                AnalyzeData(normalizedHistoricalData);
            MarketModel modeledHistoricalData =
                RunModel(analyzedHistoricalData);
            MarketRecommendation recommendation = 
                CompareModels(new[] { modeledMarketData, 
                                      modeledHistoricalData});
            return recommendation;
        }

        /// <summary>
        /// Initiates market analysis using parallel computation.
        /// </summary>        
        /// <returns>Task record that may be queried for results of the analysis</returns>
        /// <remarks>Compare with the DoAnalysisSequential method</remarks>
        public AnalysisTasks DoAnalysisParallel()
        {
            TaskFactory factory = Task.Factory;

            if (cts != null)
            {
                cts.Dispose();
                cts = null;
            }
            cts = new CancellationTokenSource();

            Task<StockDataCollection> loadNyseData =
                Task<StockDataCollection>.Factory.StartNew(
                    () => LoadNyseData(),
                    TaskCreationOptions.LongRunning);

            Task<StockDataCollection> loadNasdaqData =
                Task<StockDataCollection>.Factory.StartNew(
                    () => LoadNasdaqData(),
                    TaskCreationOptions.LongRunning);

            Task<StockDataCollection> mergeMarketData =
                factory.ContinueWhenAll<StockDataCollection, StockDataCollection>(
                    new[] { loadNyseData, loadNasdaqData },
                    (tasks) => MergeMarketData(from t in tasks select t.Result));

            Task<StockDataCollection> normalizeMarketData =
                mergeMarketData.ContinueWith(
                    (t) => NormalizeData(t.Result));

            Task<StockDataCollection> loadFedHistoricalData =
                Task<StockDataCollection>.Factory.StartNew(
                    () => LoadFedHistoricalData(),
                    TaskCreationOptions.LongRunning);

            Task<StockDataCollection> normalizeHistoricalData =
                loadFedHistoricalData.ContinueWith(
                    (t) => NormalizeData(t.Result));

            Task<StockAnalysisCollection> analyzeMarketData =
                normalizeMarketData.ContinueWith(
                    (t) => AnalyzeData(t.Result));

            Task<MarketModel> modelMarketData =
                analyzeMarketData.ContinueWith(
                    (t) => RunModel(t.Result));

            Task<StockAnalysisCollection> analyzeHistoricalData =
                normalizeHistoricalData.ContinueWith(
                    (t) => AnalyzeData(t.Result));

            Task<MarketModel> modelHistoricalData =
                analyzeHistoricalData.ContinueWith(
                    (t) => RunModel(t.Result));

            Task<MarketRecommendation> compareModels =
                factory.ContinueWhenAll<MarketModel, MarketRecommendation>(
                    new[] { modelMarketData, modelHistoricalData },
                    (tasks) => CompareModels(from t in tasks select t.Result));

            Task errorHandler = CreateErrorHandler(loadNyseData, 
                loadNasdaqData, loadFedHistoricalData, 
                mergeMarketData, normalizeHistoricalData, 
                normalizeMarketData, analyzeHistoricalData, 
                analyzeMarketData, modelHistoricalData, 
                modelMarketData, compareModels);

            return new AnalysisTasks()
                    {
                        LoadNyseData = loadNyseData,
                        LoadNasdaqData = loadNasdaqData,
                        MergeMarketData = mergeMarketData,
                        NormalizeMarketData = normalizeMarketData,
                        LoadFedHistoricalData = loadFedHistoricalData,
                        NormalizeHistoricalData = normalizeHistoricalData,
                        AnalyzeMarketData = analyzeMarketData,
                        AnalyzeHistoricalData = analyzeHistoricalData,
                        ModelMarketData = modelMarketData,
                        ModelHistoricalData = modelHistoricalData,
                        CompareModels = compareModels,
                        ErrorHandler = errorHandler
                    };
        }

        Task CreateErrorHandler(params Task[] tasks)
        {
            return Task.Factory.ContinueWhenAll(tasks, (t) =>
            {
                try
                {
                    Task.WaitAll(tasks);
                }
                catch (AggregateException e)
                {
                    Console.WriteLine(e.Flatten());
                }
            });
        }

        public void TryCancelAnalysis()
        {
            if (cts != null)
                cts.Cancel();
        }

        #endregion

        #region IDisposable Members

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        protected virtual void Dispose(bool disposing)
        {
            if (disposing)
            {
                if (cts != null)
                {
                    cts.Dispose();
                    cts = null;
                }
            }
        }

        #endregion
    }
}

File: parallel-programming-with-ms-dotnet/FSharp/Chapter5/A-Dash/Analysis/AnalysisEngine.fs
//===============================================================================
// Microsoft patterns & practices
// Parallel Programming Guide
//===============================================================================
// Copyright © Microsoft Corporation.  All rights reserved.
// This code released under the terms of the 
// Microsoft patterns & practices license (http://parallelpatterns.codeplex.com/license).
//===============================================================================

namespace Microsoft.Practices.ParallelGuideSamples.ADash

open System
open System.Collections.Generic
open System.Linq
open System.Threading
open System.Threading.Tasks
open Microsoft.Practices.ParallelGuideSamples.ADash.BusinessObjects
open Microsoft.Practices.ParallelGuideSamples.Utilities

// ------------------------------------------------------------------------------
// Implementation of the main data analysis component (contains sequential 
// version, parallel version using tasks and asynchronous workflow version)
// ------------------------------------------------------------------------------

/// Component for data analysis
/// Implements the IAnalaysisEngine interface
type AnalysisEngine(speedFactor:float) =

    let guiDispatch = Windows.Threading.Dispatcher.CurrentDispatcher
    let mutable cts : CancellationTokenSource = null

    // --------------------------------------------------------------------------
    // Various helper functions for analysis engine
    // --------------------------------------------------------------------------

    // Generate sample data for market analysis
    let generateSecurities exchange size =
      new ResizeArray<_>
        ([ for i in 0 .. size - 1 do
              yield { name = exchange + " Stock " + string i
                      priceHistory = [ 0.0; 1.0; 2.0 ] } ])

    /// Register a function as a notification to be called when task finishes
    let addNotification (task:Task<_>) f =
        task.ContinueWith
          ( new Action<Task<_>>(fun (t:Task<_>) -> f t.Result), CancellationToken.None, 
            TaskContinuationOptions.OnlyOnRanToCompletion, 
            TaskScheduler.FromCurrentSynchronizationContext()) |> ignore

    /// Trigger event on the GUI thread using a captured dispatcher
    let triggerGuiEvent (e:Event<_>) v = 
        guiDispatch.Invoke(new Action(fun () -> 
              e.Trigger(v) ), [| |]) |> ignore

    /// Constructs workflow that triggers the specified event 
    /// on the GUI thread when the wrapped async completes 
    let triggerWhenCompletes (e:Event<_>) (a:Async<_>) = async {
        let! res = a
        triggerGuiEvent e res
        return res }

    /// Utility function that creates a long running task from F# function
    let taskLong (f:unit -> 'T) = 
        Task<'T>.Factory.StartNew(new Func<'T>(f), TaskCreationOptions.LongRunning)

    // --------------------------------------------------------------------------
    // Generate sample data for analysis
    // --------------------------------------------------------------------------

    let makeNyseSecurityInfo() =
        generateSecurities "NYSE" 100

    let makeNasdaqSecurityInfo() =
        generateSecurities "NASDAQ" 100

    let makeFedSecurityInfo() =
        generateSecurities "" 100

    // Synchronous I/O intensive analysis methods

    let loadNyseData() =
        SampleUtilities.DoIoIntensiveOperation 2.5 cts.Token false |> ignore 
        if cts.Token.IsCancellationRequested then null
        else new StockDataCollection(makeNyseSecurityInfo())

    let loadNasdaqData() =
        SampleUtilities.DoIoIntensiveOperation (2.0 * speedFactor) cts.Token false |> ignore
        if cts.Token.IsCancellationRequested then null
        else new StockDataCollection(makeNasdaqSecurityInfo())

    let loadFedHistoricalData() = 
        SampleUtilities.DoIoIntensiveOperation (3.0 * speedFactor) cts.Token false |> ignore
        if cts.Token.IsCancellationRequested then null
        else new StockDataCollection(makeFedSecurityInfo())

    // Asynchronous I/O intensive analysis methods (F# specific)

    let asyncLoadNyseData() = async {
        let! _ = SampleUtilities.AsyncDoIoIntensiveOperation 2.5 cts.Token false 
        if cts.Token.IsCancellationRequested then return null
        else return new StockDataCollection(makeNyseSecurityInfo()) }

    let asyncLoadNasdaqData() = async {
        let! _ = SampleUtilities.AsyncDoIoIntensiveOperation (2.0 * speedFactor) cts.Token false 
        if cts.Token.IsCancellationRequested then return null
        else return new StockDataCollection(makeNasdaqSecurityInfo()) }

    let asyncLoadFedHistoricalData() = async {
        let! _ = SampleUtilities.AsyncDoIoIntensiveOperation (3.0 * speedFactor) cts.Token false 
        if cts.Token.IsCancellationRequested then return null
        else return new StockDataCollection(makeFedSecurityInfo()) }

    // --------------------------------------------------------------------------
    // CPU intensive analysis methods
    // --------------------------------------------------------------------------

    let mergeMarketData allMarketData =
        SampleUtilities.DoCpuIntensiveOperation (2.0 * speedFactor) cts.Token false |> ignore
        if cts.Token.IsCancellationRequested then null
        else new StockDataCollection(new ResizeArray<_>(Seq.concat allMarketData))

    let normalizeData marketData =
        SampleUtilities.DoCpuIntensiveOperation (2.0 * speedFactor) cts.Token false |> ignore
        if cts.Token.IsCancellationRequested then null
        else new StockDataCollection(marketData)

    let analyzeData data = 
        if cts.Token.IsCancellationRequested then null
        else MarketAnalyzer.Run(data)

    let runModel data =
        SampleUtilities.DoCpuIntensiveOperation (2.0 * speedFactor) cts.Token false |> ignore
        if cts.Token.IsCancellationRequested then null
        else MarketModeler.Run(data)

    let compareModels models =
        SampleUtilities.DoCpuIntensiveOperation (2.0 * speedFactor) cts.Token false |> ignore
        if cts.Token.IsCancellationRequested then null
        else ModelComparer.Run(models |> Array.ofSeq)

    let createErrorHandler tasks =
        Task.Factory.ContinueWhenAll(Array.ofList tasks, fun t ->
            try Task.WaitAll(Array.ofList tasks)
            with :? AggregateException as e ->
                Console.WriteLine(e.Flatten()) ) 

    // --------------------------------------------------------------------------
    // Implementation of the analysis engine
    // --------------------------------------------------------------------------

    new() = new AnalysisEngine(1.0)

    interface IAnalysisEngine with

        /// Creates a market recommendation using a fully sequential operation
        /// Returns a MarketRecomendation value (as the result of analysis)
        member x.DoAnalysisSequential() =
            
            // Load & process NYSE and NASDAQ data
            let nyseData = loadNyseData()
            let nasdaqData = loadNasdaqData()
            let mergedMarketData = mergeMarketData [ nyseData; nasdaqData ]
            let normalizedMarketData = normalizeData mergedMarketData
            let analyzedStockData = analyzeData normalizedMarketData
            
            // Load & process FED data
            let fedHistoricalData = loadFedHistoricalData()
            let normalizedHistoricalData = normalizeData fedHistoricalData
            let analyzedHistoricalData = analyzeData normalizedHistoricalData
            
            // Run both models and compare them
            let modeledMarketData = runModel analyzedStockData
            let modeledHistoricalData = runModel analyzedHistoricalData
            let recommendation = compareModels [ modeledMarketData; modeledHistoricalData ]
            recommendation


        /// Initiates market analysis using parallel computation (Tasks).
        /// Returns an object that can be used to register notifications with 
        /// the background calculations (and can be used to get the results)
        ///
        /// Note: Compare with the DoAnalysisSequential method
        member x.DoAnalysisParallel() =
            if cts <> null then cts.Dispose()
            cts <- new CancellationTokenSource()

            // Load & process NYSE and NASDAQ data
            let loadNyseDataTask = taskLong loadNyseData
            let loadNasdaqDataTask = taskLong loadNasdaqData

            let mergeMarketDataTask = 
              Task.Factory.ContinueWhenAll
                ( [| loadNyseDataTask; loadNasdaqDataTask |],
                  fun (tasks:Task<_>[]) -> 
                      mergeMarketData [ for t in tasks -> t.Result ])

            let normalizeMarketDataTask =
                mergeMarketDataTask.ContinueWith(fun (t:Task<_>) ->
                    normalizeData(t.Result))

            let analyzeMarketDataTask =
                normalizeMarketDataTask.ContinueWith(fun (t:Task<_>) ->
                    analyzeData(t.Result))

            // Load & process FED data
            let loadFedHistoricalDataTask = taskLong loadFedHistoricalData
            let normalizeHistoricalDataTask =
                loadFedHistoricalDataTask.ContinueWith(fun (t:Task<_>) -> 
                    normalizeData(t.Result))

            let modelMarketDataTask =
                analyzeMarketDataTask.ContinueWith(fun (t:Task<_>) ->
                    runModel(t.Result))

            let analyzeHistoricalDataTask =
                normalizeHistoricalDataTask.ContinueWith(fun (t:Task<_>) ->
                    analyzeData(t.Result))

            // Run both models and compare them
            let modelHistoricalDataTask =
                analyzeHistoricalDataTask.ContinueWith(fun (t:Task<_>) ->
                    runModel(t.Result))

            let compareModelsTask =
                Task.Factory.ContinueWhenAll
                  ( [| modelMarketDataTask; modelHistoricalDataTask |],
                    fun (tasks:Task<_>[]) -> compareModels [ for t in tasks -> t.Result ])

            // Construct error handler and return tasks as result
            let errorHandler = 
              createErrorHandler 
                [ loadNyseDataTask; loadNasdaqDataTask; loadFedHistoricalDataTask; 
                  mergeMarketDataTask; normalizeHistoricalDataTask; 
                  normalizeMarketDataTask; analyzeHistoricalDataTask; 
                  analyzeMarketDataTask; modelHistoricalDataTask; 
                  modelMarketDataTask; compareModelsTask ] 

            { new AnalysisNotifications with
                member x.AddLoadNyseData(f) = addNotification loadNyseDataTask f
                member x.AddLoadNasdaqData(f) = addNotification loadNasdaqDataTask f
                member x.AddMergeMarketData(f) = addNotification mergeMarketDataTask f
                member x.AddNormalizeMarketData(f) = addNotification normalizeMarketDataTask f
                member x.AddLoadFedHistoricalData(f) = addNotification loadFedHistoricalDataTask f
                member x.AddNormalizeHistoricalData(f) = addNotification normalizeHistoricalDataTask f
                member x.AddAnalyzeMarketData(f) = addNotification analyzeMarketDataTask f
                member x.AddAnalyzeHistoricalData(f) = addNotification analyzeHistoricalDataTask f
                member x.AddModelMarketData(f) = addNotification modelMarketDataTask f
                member x.AddModelHistoricalData(f) = addNotification modelHistoricalDataTask f
                member x.AddCompareModels(f) = addNotification compareModelsTask f
                member x.AddErrorHandler(f) = 
                    errorHandler.ContinueWith
                      ( new Action<_>(fun (t:Task<_>) -> 
                            if t.Status = TaskStatus.Faulted then f ()), 
                        TaskScheduler.FromCurrentSynchronizationContext()) |> ignore } 


        /// Initiates market analysis using asynchronous workflows (F# async).
        /// Returns an object that can be used to register notifications with 
        /// the background calculations (and can be used to get the results)
        ///
        /// To notify the caller, we create events that are triggered (on the
        /// GUI thread) when asynchronous workflow produces partial result.
        /// The returned object registers handlers with the events.
        ///
        /// Note: Compare with the DoAnalysisSequential method
        member x.DoAnalysisAsync() =
            if cts <> null then cts.Dispose()
            cts <- new CancellationTokenSource()

            // Create events that are used to notify the 
            // caller about partial results of the computation
            let loadNyseDataEvt = new Event<_>()
            let loadNasdaqDataEvt = new Event<_>()
            let mergeMarketDataEvt = new Event<_>()
            let normalizeMarketDataEvt = new Event<_>()
            let loadFedHistoricalDataEvt = new Event<_>()
            let normalizeHistoricalDataEvt = new Event<_>()
            let analyzeMarketDataEvt = new Event<_>()
            let analyzeHistoricalDataEvt = new Event<_>()
            let modelMarketDataEvt = new Event<_>()
            let modelHistoricalDataEvt = new Event<_>()
            let compareModelsEvt = new Event<_>()
            let errorHandlerEvt = new Event<_>()

            // Load & process NYSE and NASDAQ data
            let marketModel = async {
                // Start loading of data from two sources in background (both I/O
                // operations are perfomed asynchronously without blocking threads)
                let! nyse = 
                  asyncLoadNyseData() 
                  |> triggerWhenCompletes loadNyseDataEvt |> Async.StartChild
                let! nasdaq = 
                  asyncLoadNasdaqData() 
                  |> triggerWhenCompletes loadNasdaqDataEvt |> Async.StartChild
                
                // Wait for both tasks to complete and continue
                let! nyseData = nyse
                let! nasdaqData = nasdaq
                let merged = mergeMarketData [ nyseData; nasdaqData ] 
                triggerGuiEvent mergeMarketDataEvt merged
                
                // Perform analysis of the merged data
                let normalized = normalizeData merged
                triggerGuiEvent normalizeMarketDataEvt normalized
                let analyzed = analyzeData normalized 
                triggerGuiEvent analyzeMarketDataEvt analyzed
                let res = runModel analyzed 
                triggerGuiEvent modelMarketDataEvt res
                return res }

            // Load & process FED data
            let historicalModel = async {
                // Obtain data asynchronously using non-blocking I/O
                let! fed = asyncLoadFedHistoricalData()
                triggerGuiEvent loadFedHistoricalDataEvt fed

                // Perform CPU-intensive analysis of the data
                let normalized = normalizeData fed
                triggerGuiEvent normalizeHistoricalDataEvt normalized
                let analyzed = analyzeData normalized
                triggerGuiEvent analyzeHistoricalDataEvt analyzed
                let res = runModel analyzed 
                triggerGuiEvent modelHistoricalDataEvt res
                return res }

            // Run both of the models and compare them to get recommendation
            // When error occurs, it is propagated here and handled using 'try'
            let compare = async {
                try
                    let! models = Async.Parallel [ marketModel; historicalModel ]
                    let res = compareModels models 
                    triggerGuiEvent compareModelsEvt res 
                with e ->
                    triggerGuiEvent errorHandlerEvt ()  }

            // Construct error handler and return tasks as result
            // (if we pass cancellation token to 'Async.Start', the workflow
            // will be terminated automatically, so we wouldn't return 'null'
            // to the caller - which is expected in the current design)
            Async.Start(compare)

            { new AnalysisNotifications with
                member x.AddLoadNyseData(f) = loadNyseDataEvt.Publish.Add(f)
                member x.AddLoadNasdaqData(f) = loadNasdaqDataEvt.Publish.Add(f)
                member x.AddMergeMarketData(f) = mergeMarketDataEvt.Publish.Add(f)
                member x.AddNormalizeMarketData(f) = normalizeMarketDataEvt.Publish.Add(f)
                member x.AddLoadFedHistoricalData(f) = loadFedHistoricalDataEvt.Publish.Add(f)
                member x.AddNormalizeHistoricalData(f) = normalizeHistoricalDataEvt.Publish.Add(f)
                member x.AddAnalyzeMarketData(f) = analyzeMarketDataEvt.Publish.Add(f)
                member x.AddAnalyzeHistoricalData(f) = analyzeHistoricalDataEvt.Publish.Add(f)
                member x.AddModelMarketData(f) = modelMarketDataEvt.Publish.Add(f)
                member x.AddModelHistoricalData(f) = modelHistoricalDataEvt.Publish.Add(f)
                member x.AddCompareModels(f) = compareModelsEvt.Publish.Add(f)
                member x.AddErrorHandler(f) = errorHandlerEvt.Publish.Add(f) } 

        /// Cancel the running parallel or asynchronous workflow based
        /// analysis (this causes the analysis to return 'null' as the result)
        member x.TryCancelAnalysis() =
            if cts <> null then cts.Cancel()

    // --------------------------------------------------------------------------
    // Implementation of the disposable pattern
    // --------------------------------------------------------------------------
    
    member x.Dispose(disposing) =
        if disposing then
            if cts <> null then
                cts.Dispose()
                cts <- null

    interface IDisposable with 
        member x.Dispose() =
            x.Dispose(true)
            GC.SuppressFinalize(x)
